CMakeLists.txt

```diff
@@ -797,7 +797,7 @@ elseif("${SWIFT_HOST_VARIANT_SDK}" MATCHES "(OSX|IOS*|TVOS*|WATCHOS*)")
   #
   # Primary variant is always OSX; even on iOS hosts.
   set(SWIFT_PRIMARY_VARIANT_SDK_default "OSX")
-  set(SWIFT_PRIMARY_VARIANT_ARCH_default "${CMAKE_HOST_SYSTEM_PROCESSOR}")
+  set(SWIFT_PRIMARY_VARIANT_ARCH_default "${CMAKE_SYSTEM_PROCESSOR}")
```

Comment:
I think that the primary variant being the host is right - that is needed for running the tests (which use the primary variant).  Now, arguably that would mean that you always need to build for the host even if you don't need it.

Reply:
It is true that `check-*` (without any platform suffix) depends on the primary variant, and that's the problem.

`CMAKE_SYSTEM_PROCESSOR` is used to find the `SWIFT_HOST_VARIANT_ARCH`, but `CMAKE_HOST_SYSTEM_PROCESSOR` is used to find the `SWIFT_PRIMARY_VARIANT_ARCH`. If one sets `CMAKE_SYSTEM_PROCESSOR`, the tools will target the right arch, but the primary variant will be off (so stdlib and tests).

BTW, this doesn't change the standard flow of building for the host machine, or using `CMAKE_OSX_ARCHITECTURES` to build universal binaries, only when trying to cross-compile from Intel to Arm64.

Reply:
This gets confusing, but the primary variant arch should be the host arch by default. An alternative would be to use @davidungar 's approach of setting it to `SWIFT_HOST_VARIANT_ARCH` instead, #35231, [which is what every other Swift platform does](https://github.com/apple/swift/blob/0750dde43c270ec6de6ee219eaea69decd8a4f9a/CMakeLists.txt#L705). If you're cross-compiling the stdlib, you could override the primary variant, [as shown in the iOS build preset](https://github.com/apple/swift/blob/44708dc739fe19aaeafc4360bed5682c12b032d0/utils/build-presets.ini#L2255), though that shouldn't be required.

I think using the same variable as every other platform as in #35231 would appear to be the best, but [I notice that may use a different arch name, ie ` arm64` and not `arm64e`](https://github.com/apple/swift/blob/0750dde43c270ec6de6ee219eaea69decd8a4f9a/CMakeLists.txt#L630), and I don't know what difference that makes.

Reply:
The trick of using `SWIFT_HOST_VARIANT_ARCH` might not work if the `SWIFT_HOST_VARIANT_SDK` is not `OSX`. The comment on top of this lines say "Primary variant is always OSX; even on iOS hosts.". If `IOS*|TVOS*|WATCHOS*` is the host variant, the architecture might be wrong for the host (it could be `arm64_32` or something weirder).

What I am trying to cross-compile is both the compiler/tools, and the build system also builds the stdlib. A beefy Intel Mac compiles, and a fairly-good-but-not-as-beefy Arm64 Mac executes. Without this change, the primary variant is selected as the host machine, which set ups the targets depending on the x86_64, but only configures the arm64 ones, creating a problem.

It is true that I can simply overfit the setup and pass my own defaults for `SWIFT_PRIMARY/HOST_VARIANT_SDK/ARCH` and that should work, but I hoped that simply switching `CMAKE_SYSTEM_PROCESSOR` (the standard way of doing things in CMake) was not going to be a problem. If it is, I am happy to remove the diff from review.

Can we all agree at least that the Swift build setup is very inadequate with respect to cross-compilation? I think everybody in this conversation would agree :D

Reply:
> The trick of using SWIFT_HOST_VARIANT_ARCH might not work if the SWIFT_HOST_VARIANT_SDK is not OSX. The comment on top of this lines say "Primary variant is always OSX; even on iOS hosts.". If IOS*|TVOS*|WATCHOS* is the host variant, the architecture might be wrong for the host (it could be arm64_32 or something weirder).

There are only two ways this can happen:

- You are building the Swift toolchain directly on an iPhone/iPad/etc. non-macOS host, which is almost impossible and unsupported.

- You are cross-compiling the toolchain for non-macOS, in which case, you're expected to pass in the primary variant arch as I showed in the iOS build preset I linked above.

So those scenarios are unlikely and this variable won't fail because of that.

> Without this change, the primary variant is selected as the host machine, which set ups the targets depending on the x86_64, but only configures the arm64 ones, creating a problem.

Understood, but setting `SWIFT_HOST_VARIANT_ARCH` as the default here should also fix that.

> It is true that I can simply overfit the setup and pass my own defaults for SWIFT_PRIMARY/HOST_VARIANT_SDK/ARCH and that should work, but I hoped that simply switching CMAKE_SYSTEM_PROCESSOR (the standard way of doing things in CMake) was not going to be a problem.

There will be no need to pass in _any_ of those CMake variables on the command-line if this default is chosen well.

> Can we all agree at least that the Swift build setup is very inadequate with respect to cross-compilation?

Oh yes, I've been complaining since my first pull two years ago about these redundant and often interchangeable CMake variables.

Two things you should check before we move forward:

- What is the `CMAKE_HOST_SYSTEM_PROCESSOR` detected on an M1 Mac, `arm64` or `arm64e`, because `SWIFT_HOST_VARIANT_ARCH` chops off the trailing `e` on macOS? Does that make a difference?

- The macOS CI currently cross-compiles the toolchain for Mac arm64 on every CI run. What preset is it using to do this and what different flags is it using than you?

Reply:
> What is the CMAKE_HOST_SYSTEM_PROCESSOR detected on an M1 Mac, arm64 or arm64e, because SWIFT_HOST_VARIANT_ARCH chops off the trailing e on macOS? Does that make a difference?

It reports `arm64` (in a completely empty CMake project, with a native arm64 CMake). I imagine that in macOS `arm64` is probably equivalent to `arm64e`, but I don't know why the name is different. Probably some Apple folks can answer better.

> The macOS CI currently cross-compiles the toolchain for Mac arm64 on every CI run. What preset is it using to do this and what different flags is it using than you?

The `build-script-impl` has a final lipo stage using `utils/lipo-recursive` that merges the two different archs into universal binaries (the two compilations seems to do double work with at least pieces of the stdlib, btw).

I don't use the build-script. I was trying to cross-compile from one arch to the other, before trying universal (by using `CMAKE_OSX_ARCHITECTURES`.

Reply:
> It reports arm64 (in a completely empty CMake project, with a native arm64 CMake).

Good, that means using `SWIFT_HOST_VARIANT_ARCH` here will make no difference when natively compiling on Mac arm64, but will fix the cross-compilation scenario you're having problems with.

> I don't use the build-script.

OK, this explains a lot, as that passes in a lot of needed CMake variables by default.